### DISCUSSION TITLE
gnark solidity check filed due to ExportSolidity Public Input Size

### DISCUSSION THREAD

**Discussion by zhengjunhe** [Consensys/gnark#893] [Q&A] [answered]
Hi
I develop a circuit and do the test via 
func (assert *Assert) ProverSucceeded(circuit frontend.Circuit, validAssignment frontend.Circuit, opts ...TestingOption) 
but it failed due to gnark solidity check as ExportSolidity Public Input Size is one more and it make check fail as below
	if nbInputs != nbPublicInputs {
		panic("nbInputs != nbPublicInputs")
	}
nbInputs is always nbInputs = nbPublicInputs - 1
I searched the similar problem in the discussion and found https://github.com/Consensys/gnark/discussions/644. 

How to solove this problem?
Is there any skill to write circuit make nbPublicInputs same as nbInputs?

**Comment by ivokub** [comment]
Hi, it is a known issue for Groth16 verifier when using `api.Commit` (which we use for field emulation). See more here: https://github.com/Consensys/gnark/issues/860

Right now the workaround is to use PLONK verifier.

**Reply by zhengjunhe** [reply]
@ivokub,I try the PLONK verifier by call test function
func (assert *Assert) ProverSucceeded(circuit frontend.Circuit, validAssignment frontend.Circuit, opts ...TestingOption) {
and set the following two flag to true to enable do the gnark-solidity-checker.
releaseTestFlag = true
solcCheck = true
But,unfortunately，the gnark-solidity-checker failed with plonk.
I checked that the proof can be verified locally.
Could you pls give me some tips to work out this problem.

Below is some info releated to the test:
        	Test:       	TestEth1BlockHashProofs_BN254/bn254/plonk/valid_witness/solidity
        	Messages:   	verify called
        	            	generating /tmp/gnark-solidity-check1573857423/main.go
        	            	generating /tmp/gnark-solidity-check1573857423/go.mod
        	            	running go mod tidy: /usr/local/go/bin/go mod tidy
        	            	running go run: /usr/local/go/bin/go run main.go gnark_solidity.go
        	            	error: exit status 1
        	            	output: proof is invalid
        	            	exit status 42
        	            	
                --- FAIL: TestEth1BlockHashProofs_BN254/bn254/plonk/valid_witness/solidity (1.24s)

=== RUN   TestEth1BlockHashProofs_BN254/bn254/plonk/valid_witness/proof
                --- PASS: TestEth1BlockHashProofs_BN254/bn254/plonk/valid_witness/proof (0.00s)

**Reply by zhengjunhe** [reply]
I find my circuit can't pass the solidity checker may caused by my function call via:
gnark/std/hash/sha2.
Which is uesed for calculating sha2-hash.

I also do the sha2 unit test to make sure whether it could pass the solidity checker via the adding below:
Step 1: adding the below to function:TestSHA2
	garkAssert := test.NewAssert(t)
	garkAssert.ProverSucceeded(&sha2Circuit{In: make([]uints.U8, len(bts))}, &witness,
		test.WithCurves(ecc.BN254),
		test.WithCompileOpts(frontend.IgnoreUnconstrainedInputs()),
		test.WithBackends(backend.PLONK)) 
Step 2: declare the Expected to be public
type sha2Circuit struct {
	In       []uints.U8
	Expected [32]uints.U8 `gnark:",public"`
}
Step 3:enable solidity check
releaseTestFlag = true
solcCheck = true

@ivokub,Could you pls give some comments.
Thanks

**Comment by zhengjunhe** [comment]
Hi，could you pls check the problem I reported here:
https://github.com/Consensys/gnark/discussions/893.
Thanks.

Ivo Kubjas ***@***.***> 于2023年10月30日周一 19:59写道：

**Reply by ivokub** [reply]
Could you post a minimal working example properly formatted? I could try debugging if I can copy-paste the program, but won't have time to implement the circuits, witness generation etc. Thanks.

**Reply by zhengjunhe** [reply]
import (
	"crypto/sha256"
	"fmt"
	"github.com/consensys/gnark/std/hash/sha2"
	"github.com/consensys/gnark/backend"
	"testing"

	"github.com/consensys/gnark-crypto/ecc"
	"github.com/consensys/gnark/frontend"
	"github.com/consensys/gnark/std/math/uints"
	"github.com/consensys/gnark/test"
)

type sha2Circuit struct {
	In       []uints.U8
	Expected [32]uints.U8 `gnark:",public"`
}

func (c *sha2Circuit) Define(api frontend.API) error {
	h, err := sha2.New(api)
	if err != nil {
		return err
	}
	uapi, err := uints.New[uints.U32](api)
	if err != nil {
		return err
	}
	h.Write(c.In)
	res := h.Sum()
	if len(res) != 32 {
		return fmt.Errorf("not 32 bytes")
	}
	for i := range c.Expected {
		uapi.ByteAssertEq(c.Expected[i], res[i])
	}
	return nil
}

func TestSHA2(t *testing.T) {
	bts := make([]byte, 310)
	dgst := sha256.Sum256(bts)
	witness := sha2Circuit{
		In: uints.NewU8Array(bts),
	}
	copy(witness.Expected[:], uints.NewU8Array(dgst[:]))
	err := test.IsSolved(&sha2Circuit{In: make([]uints.U8, len(bts))}, &witness, ecc.BN254.ScalarField())
	if err != nil {
		t.Fatal(err)
	}

	garkAssert := test.NewAssert(t)
	garkAssert.ProverSucceeded(&sha2Circuit{In: make([]uints.U8, len(bts))}, &witness,
		test.WithCurves(ecc.BN254),
		test.WithCompileOpts(frontend.IgnoreUnconstrainedInputs()),
		test.WithBackends(backend.PLONK))
}

1.set the following 2 vals to true in gnark

const SolcCheck = true
releaseTestFlag = true

2.run the unit test:
TestSHA2

**Reply by zhengjunhe** [reply]
@ivokub, pls copy the code above and run the unit test, thanks

**Comment by zhengjunhe** [comment]
Sure, no problem.

Ivo Kubjas ***@***.***>于2023年11月8日 周三18:16写道：

**Comment by zhengjunhe** [comment]
I think you forked gnark,can I push a branch to your rep.And then you could
see my example?Is it ok?

Ivo Kubjas ***@***.***>于2023年11月8日 周三18:16写道：

**Comment by zhengjunhe** [comment]
https://github.com/Consensys/gnark/pull/907

Is it ok for you?

Zhengjun He ***@***.***> 于2023年11月8日周三 18:32写道：

**Comment by ivokub** [comment]
@zhengjunhe - I tested on my side and it works. Do you have the application `gnark-solidity-checker` installed? See the repository https://github.com/consensys/gnark-solidity-checker for dependency installation instructions. You can install the application by `go install github.com/consensys/gnark-solidity-checker`, but keep in mind that `$GOPATH/bin` has to be defined in `$PATH` so that the command would be found.

Additionally, solidity checker requires that there are public inputs. In the SHA2 circuit in gnark all the inputs are private. Change:
```go
type sha2Circuit struct {
	In       []uints.U8
	Expected [32]uints.U8
}
```
to
```go
type sha2Circuit struct {
	In       []uints.U8 `gnark:",public"`
	Expected [32]uints.U8
}
```

Finally, to run the tests without having to modify the files, use the command:
```bash
go test -timeout 10m -tags debug,solccheck,prover_checks -run ^TestSHA2$ github.com/consensys/gnark/std/hash/sha2 -v -count=1
```

Ultimately - PLONK verifier works as intended. The tests are meant for being used in continuous integration tests, not directly by the end users.

**Reply by ivokub** [reply]
To wrap up - I'll close this discussion for now. Tests work and the overhead to debug your installation is too big.

**Reply by zhengjunhe** [reply, marked as answer]
ok,thanks.
